packages/libandroid-posix-semaphore/semaphore.c

```diff
@@ -0,0 +1,244 @@
+/* This file is a port of posix named semaphore for Termux Android, 
+   based on musl-libc which is licensed under the standard MIT license. 
+   The ported files are listed as following.
+
+   File(s): src/thread/sem_open.c
+            src/thread/sem_unlink.c
+
+   Copyright © 2005-2020 Rich Felker, et al.
+
+   Permission is hereby granted, free of charge, to any person obtaining
+   a copy of this software and associated documentation files (the
+   "Software"), to deal in the Software without restriction, including
+   without limitation the rights to use, copy, modify, merge, publish,
+   distribute, sublicense, and/or sell copies of the Software, and to
+   permit persons to whom the Software is furnished to do so, subject to
+   the following conditions:
+
+   The above copyright notice and this permission notice shall be
+   included in all copies or substantial portions of the Software.
+
+   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+   EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+   MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+   IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+   CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+   TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+   SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include "semaphore.h" // sem_t, sem_init()
+#include <sys/mman.h>  // mmap(), munmap()
+#include <limits.h>    // SEM_NSEMS_MAX, NAME_MAX, INT_MAX
+#include <fcntl.h>     // open()
+#include <unistd.h>    // write(), close(), unlink()
+#include <string.h>    // strlen(), memcpy()
+#include <stdarg.h>    // va_arg
+#include <errno.h>     // errno
+#include <sys/stat.h>  // fstat()
+#include <stdlib.h>    // calloc()
+#include <pthread.h>   // mutex
+
+#ifndef SEM_NSEMS_MAX
+#define SEM_NSEMS_MAX 256
+#endif // !SEM_NSEMS_MAX
+
+static __inline__ char *__strchrnul(const char *s, int c)
+{
+    c = (unsigned char)c;
+    if (!c) return (char *)s + strlen(s);
+    for (; *s && *(unsigned char *)s != c; s++);
+    return (char *)s;
+}
+
+typedef struct {
+    ino_t ino;
+    sem_t *sem;
+    int refcnt;
+} semtab_type;
+
+static semtab_type *semtab;
+static pthread_mutex_t lock;
+
+#define LOCK(l) (pthread_mutex_lock(&l))
+#define UNLOCK(l) (pthread_mutex_unlock(&l))
+
+#define FLAGS (O_RDWR|O_NOFOLLOW|O_CLOEXEC|O_NONBLOCK)
+
+static char *__sem_mapname(const char *name, char *buf)
+{
+    char *p;
+    while (*name == '/') name++;
+    if (*(p = __strchrnul(name, '/')) || p==name ||
+        (p-name <= 2 && name[0]=='.' && p[-1]=='.')) {
+        errno = EINVAL;
+        return 0;
+    }
+    if (p-name > NAME_MAX-4) {
+        errno = ENAMETOOLONG;
+        return 0;
+    }
+    memcpy(buf, "/data/data/com.termux/files/usr/tmp/sem.", 40);
```

Comment:
Can we avoid hardcode prefix? Preferably use @TERMUX_PREFIX@

Reply:
This is a C source file, not a patch. Seems the shortcut `@TERMUX_PREFIX@` can only work on a .patch file.

Reply:
For non-patch files, we have to `sed` manually to substitute `@TERMUX_PREFIX@`:

https://github.com/termux/termux-packages/blob/e1a6c93beac65b01267b93506b6b2e5826208281/packages/liblua54/build.sh#L15-L17

This is preferable, but not mandatory I suppose. We have another instance where `$TERMUX_PREFIX` is hardcoded:

https://github.com/termux/termux-packages/blob/e1a6c93beac65b01267b93506b6b2e5826208281/packages/libandroid-wordexp/wordexp.c#L121

Reply:
Checking on wordexp.c seems like xeffyr was fine with full prefix. Guess this is a go then.

Reply:
It is preferable avoid hardcoding it, sed'ing it would be better